test/replication-luatest/gh_7590_replicaset_assert_test.lua

```diff
@@ -0,0 +1,91 @@
+local t = require('luatest')
+local server = require('luatest.server')
+local cluster = require('luatest.replica_set')
+
+local g = t.group('gh-7590')
+
+g.before_all(function(g)
+    g.cluster = cluster:new({})
+    local replication_cfg = {
+        server.build_listen_uri('r1'),
+        server.build_listen_uri('r2'),
+        server.build_listen_uri('r3'),
+    }
+
+    g.r1 = g.cluster:build_and_add_server({alias = 'r1', box_cfg = {
+        -- We need to specify uuid explicitly, so that rb tree traversal
+        -- wouldn't take errored applier (applier from r2 to r3) first.
+        instance_uuid = '00000000-0000-0000-0000-000000000001',
+        replication = replication_cfg,
+    }})
+
+    g.r2 = g.cluster:build_and_add_server({alias = 'r2', box_cfg = {
+        instance_uuid = '00000000-0000-0000-0000-000000000002',
+        replication = replication_cfg,
+    }})
+    g.r3 = g.cluster:build_and_add_server({alias = 'r3', box_cfg = {
+        instance_uuid = '00000000-0000-0000-0000-000000000003',
+        replication = replication_cfg,
+    }})
+
+    g.cluster:start()
+    g.r2:exec(function()
+        box.schema.create_space('test'):create_index('pk')
+    end)
+
+    g.cluster:wait_for_fullmesh()
+    g.r3:wait_for_vclock_of(g.r2)
+end)
+
+g.after_all(function(g)
+    g.cluster:drop()
+end)
+
+g.test_replicaset_state_machine_assert_fail = function(g)
+    g.r1:exec(function()
+        -- Forbid r1 to send any kind of data in order to avoid
+        -- triggering destruction of the applier from r1 to r3.
+        box.error.injection.set('ERRINJ_RELAY_SEND_DELAY', true)
```

Comment:
The test still hangs occasionally when run for 256 times in 32 jobs.
I think the reason is that RELAY_SEND_DELAY forbids relay to send heartbeats, and sometimes r2 destroys applier due to time out.

My first idea was to replace `ERRINJ_RELAY_SEND_DELAY` with `ERRINJ_WAL_DELAY`. This way relay continues to write heartbeats, but doesn't send any actual data, because no data can be written on r1.
But this would lead to timeouts between `r1` and `r2`.

Let's simply reconfigure r1 to not replicate from r2 after bootstrap. No error injections needed.

Reply:
P.S. even after trying ERRINJ_WAL_DELAY and reconfiguring replication the test hangs for a couple out of 256 runs. Please, check it out on your side.

Reply:
Tried to reproduce hang locally on my laptop and desktop. Unfortunately, it doesn't hang at all, no matter, how many tests are run or how many threads are used. Don't have access to dev machines now too(